Update ONNX's IO Adapter to support FakeTensor with ExportedProgram

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115380
* #115353
* #114762
* #115281
* __->__ #114407

Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

ps: https://github.com/pytorch/pytorch/issues/115461 will track the API revision for the cases where additional `model_with_state_dict` are required to produce complete ONNX files exported with fake support. This is also tracked by the umbrella fake tensor issue https://github.com/pytorch/pytorch/issues/105464 FYI @BowenBao

Commit:
Update ONNX's IO Adapter to support FakeTensor with ExportedProgram

Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -681,21 +681,26 @@ def __init__(
-        self, *args: Any, options: Optional[ONNXRuntimeOptions] = None, **kwargs: Any
+        self,
+        *args: Any,
+        model: Union[torch.nn.Module, Callable, torch_export.ExportedProgram],
```
I think we can design an api to load concrete weights for ONNXProgram (Concrefy - as opposed to fakefy). And make it lazy loading.

Both `__call__`, and `save` benefit from it.

thiagocrepaldi replied:
Yup, I believe we had discussed this not long ago.

That will be especially useful when fake mode is used to export super large models. the weight-by-weight loading from a checkpoint will be paramount to create an ONNX file with memory constraints.

I will create a dedicated PR for this; that might include a different format file if we cannot partially load conventional torch checkpoints

BowenBao replied:
I have two main concerns.

1. There is implicitly a "fake" version of `ONNXProgram`, where initializers, buffers etc are not filled in model_proto. This PR is on the direction of solidifying that concept. It is better to call it out explicitly and design apis around it.

2. I feel like the massive change of api is going backwards ... I feel it is mixing two use cases 1) fake export/ training where weights are not part of proto 2) aot export where constants are lifted and not part of proto. Is this a temporary change or do you plan to change it back in future dedicated PR?

BowenBao replied:
I don't remember why we are not filling in the lifted constants by aot_export into model_proto prior to this step. Could you elaborate?

thiagocrepaldi replied:
1. I didn't fully understand what you meant by "implicitly fake ONNXProgram". IIUC, `ONNXProgram` while in memory doesn't have weights (by design of fake tensor). But the `ONNXProgram.save` already fills the serialized ONNX file with initializers since when we implemented fake tensor support for the `torch._dynamo.export` fx tracer.

2. This PR does handle 1) fake tensor with "external state_dict" to call the exported model in memory (as the previous comment explained, `ONNXProgram.save` already was able to save a complete ONNX model with initializers) and 2) constants being lifted by AOTAutograd. 1) was the goal of the PR (fake tensor) and 2) ended up a side effect to enable a fake tensor test using google t5 or openai whisper (they had constants lifted). As we follow the resulting GraphModule as is and it has lifted constants, the resulting onnx model had the constant lifted. I have no strong preference over this; just made the smallest change to make it work

thiagocrepaldi replied:
> for 2. I saw the next PR changed it a bit, it does feel like we will be constantly changing it so this is more like temporary solution?

I am doing changes incrementally as the PRs are growing and the merge process is taking quite a bit to happen :)

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/torch_export_graph_extractor.py:
```diff
@@ -48,21 +48,29 @@ def generate_fx(
-        updated_model_args = self.input_adapter.apply(*model_args, **model_kwargs)
+        # NOTE: temp workaround for https://github.com/pytorch/pytorch/issues/99534
+        # Dynamo doesn't support non-tensor inputs.
+        options.fx_tracer.input_adapter.append_step(
```
[Borrowed](https://github.com/pytorch/pytorch/pull/114886/files#diff-adfa00757e9e65b2e63fe6f38c600c2edea8be7931929a45b0402ba5befca52dR58-R62) from #114886

titaiwangms replied:
Thanks!

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/exporter.py:
```diff
@@ -851,18 +860,20 @@ def adapt_torch_inputs_to_onnx(
     def adapt_torch_outputs_to_onnx(
-        self, model_outputs: Any
+        self,
+        model: Union[torch.nn.Module, Callable, torch_export.ExportedProgram],
```
For my education, in `ExportedProgram`, when it lifts the buffers and constants, do they maintain a `state_dict` to store them as implicit inputs for users? If they do, it should be accessible from `ExportedProgram`, otherwise, with the same args and kwargs, we can't run it. So, I am wondering why we are moving args "model" from traced to user provided during adapt inputs/outpus. Shouldn't the original traced model have everything?

thiagocrepaldi replied:
ExportedProgram does have `state_dict`, but they are fakefied when fake tensor is enabled during export. For this scenario (only), we need an external (non-fake) state_dict.

The next PR will make `model` optional to be used only for fake tensor scenario

titaiwangms replied:
Ah right. Only if we could enable fake mode inside `torch.onnx.dynamo_export`, we can avoid have users provide models twice during export... Do you think it's doable?

thiagocrepaldi replied:
This way, they model would have to fit in memory to be passed on to `dynamo_export`. Having it externally might open us a way to convert the state_dict directly to onnx external data. This is what I am going to work on after this series of PRs for hte "basic" fake tensor support

titaiwangms replied:
Make sense. I forgot that.

titaiwangms replied:
Sorry, just one another Q. So if the model is not loadable or openable, how do we expect users to pass in during adapt inputs/outputs. Where is the crucial stage we facing memory issue?

thiagocrepaldi replied:
Fake support has 2 stages. 

1. export the model, which is the goal of this PR. For testing, we are assuming the model fits in memory (in a temporary basis)
2. Convert/load a pytorch state_dict/checkpoint into the ONNX model. This may change how input/output adapt works

You questions seems to be the second stage, which we are not tackling just yet, That will be the next task, which might involve serializing a regular pytorch checkpoint directly into ONNX and then merge it with the ONNX file. This way, `adapt_torch_*` API wouldn't need the `model_with_state_dict`

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/serialization.py:
```diff
@@ -75,7 +75,7 @@ def _create_tensor_proto_with_external_data(
         # data_file.seek(0)
         # Write tensor content to the file.
-        data_file.write(tensor.numpy().tobytes())
+        data_file.write(tensor.numpy(force=True).tobytes())
```
when the tensor has `requires_grad=True` set we need to force. The "old" way would be calling `.detach()`, but `force=True` is preferred

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/pytorch_test_common.py:
```diff
@@ -188,11 +188,13 @@ def wrapper(self, *args, **kwargs):
 
 
-def skip_dynamic_fx_test(reason: str):
+def skip_dynamic_fx_test(reason: str, skip_model_type=None):
```
It was on "on purpose" to prevent circular dependency between `pytorch_test_common.py` and `onnx_test_common.py`

thiagocrepaldi commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -1035,6 +1081,10 @@ def _test_fake_tensor_mode_exporter(
+    @pytorch_test_common.skip_dynamic_fx_test(
+        "AssertionError: Dynamic shape check failed for graph inputs",
+        skip_model_type=onnx_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM,
+    )
```
This is exactly what `@pytorch_test_common.skip_dynamic_fx_test(skip_model_type=onnx_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM)` does. It only skip dynamic shape for `skip_model_type` model type. e.g. `torch.nn.Module` model type will run with dynamic shape

Commit:
Update on "Update ONNX's IO Adapter to support FakeTensor with ExportedProgram"


Currently, the ONNX exporter using torch.nn.Module as input can support
FakeTensor because the ONNX model stores all initializers

When using torch.export.ExportedProgram as input, the initializers are
lifted as inputs. In order to execute the ONNX model, we need to pass a
reference to the non-fake model to the
ONNXProgram.adapt_torch_inputs_to_onnx API, so that initializers can be
fetched from the model and fed to the ONNX model as input

[ghstack-poisoned]

titaiwangms reviewed (commented):
Sorry, I forgot to mention, there are large scale models relying on fake_mode in `test_fx_to_onnx.py`, which are exportable, but not runnable with ORT due to non-persistent buffers. It's good opportunity to bring them back to runtime with `ExportedProgram`.

https://github.com/pytorch/pytorch/blob/686a3e0bf08af50a3b83532058fc5a6667365e51/test/onnx/test_fx_to_onnx.py#L531-L632

BowenBao reviewed (approved):
🚢 to unblock CI and experiments. Let's keep exploring and revisit https://github.com/pytorch/pytorch/issues/115461

thiagocrepaldi commented:
@pytorchbot merge
